### PR TITLE
[Draft] Improved tree shaking code

### PR DESCRIPTION
**Pull request by EloB**
# Draft

This is a draft to improve tree shaking and dead code elimination when bundling `typia`.

I created this PR to get early feedback.

When I have more time I will provide an example how I use this combined with a esbuild plugin to bundle typia automatically and that it tree shakes everything that it doesn't need.

## Code changes:
- Replaced `namespace name {}` with `export * as namespaceName from "./name"`. Improves dead code elimination.
- Added `/** @__PURE__ */` to some exports that has `Object.assign`.

## Example
I bundled the code below with esbuild and I was able to compress the bundle from 69827 characters down to 5804 (**-92%**). We also need to add `"sideEffects": false` to `ret` and `randexp` package.json files. I've made a PR to `randexp` but have to make another for `ret`.
```ts
import { IValidation, tags, validate } from "typia";

const res: IValidation<IMember> = validate<IMember>({
  id: 5, // wrong, must be string (uuid)
  age: 20.75, // wrong, not integer
  email: "samchon.github@gmail.com",
});

if (!res.success) console.log(res.errors);

interface IMember {
  id: string & tags.Format<"uuid">;
  email: string & tags.Format<"email">;
  age: number
  & tags.Type<"uint32">
  & tags.ExclusiveMinimum<19>
  & tags.Maximum<100>;
}
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
   "scripts": {
     "test": "npm run package:tgz",
     "-------------------------------------------------": "",
-    "build": "rimraf lib && tsc --removeComments --declaration false && tsc --emitDeclarationOnly",
+    "build": "rimraf lib && tsc --declaration false && tsc --emitDeclarationOnly",
```
Comment (EloB) [inline package.json:18]: I include the comments so it doesn't strip `/** @__PURE__ */`.

**Comment by samchon** [comment]
Thanks for contribution, but bundled size is increased from 11.2kb to 11.5kb.

  - https://bundlephobia.com/package/typia@5.3.11
  - https://bundlephobia.com/package/typia@5.3.12-dev.20240118

However, it seems like you're expert at the package bundling. I'm considering another bundling diet way through `require()` function. What about changing `typia` to generate `require()` function instead of using the namespace like below? If this way works, it may possible to diet near to the 0 byte.

```typescript
// CURRENT
const $guard = typia_1.default.json.createAssertStringify.guard;

// HOW ABOUT THIS ONE?
const $guard = require("typia/lib/functional/$guard").$guard;
```

**Review by samchon — COMMENTED** [review]
Thanks for contribution, but bundled size is increased from 11.2kb to 11.5kb.

  - https://bundlephobia.com/package/typia@5.3.11
  - https://bundlephobia.com/package/typia@5.3.12-dev.20240118

However, it seems like you're expert at the package bundling. Looking your PR, I got another idea of bundling diet way through the `require()` function. What about changing `typia` to generate `require()` function instead of using the namespace like below? 

If this way works, it may possible to diet near to the 0 byte, so that don't need to consider tree shaking problem.

```typescript
// CURRENT
const $guard = typia_1.default.json.createAssertStringify.guard;

// HOW ABOUT THIS ONE?
const $guard = require("typia/lib/functional/$guard").$guard;
```

**Comment by EloB** [comment]
@samchon That is a even better improvement but requires a lot of rewriting. Maybe take small steps? Make this happen then gradually moving towards that? 🙂

The npm package might got bigger from my changes but when you try to bundle this as a dependency into a single file using tools like esbuild, webpack and rollup than you will see massive improvements. I need this when I write AWS lambda functions to improve performance and keep the file small by dead code elimination.

I think you still need to consider tree shaking when you call functions in the root scope by for instance `Object.assign`.

File: src/functional/$ProtobufReader.ts [hunk 2]
```diff
@@ -189,7 +189,7 @@ export class $ProtobufReader {
   }
 }
 
-const utf8 = new TextDecoder();
-const N01 = BigInt(0x01);
-const N7F = BigInt(0x7f);
-const N80 = BigInt(0x80);
+const utf8 = /** @__PURE__ */ new TextDecoder();
+const N01 = /** @__PURE__ */ BigInt(0x01);
```
Comment (EloB) [inline src/functional/$ProtobufReader.ts:193]: I think these could be change to not execute functions in the root scope.

```js
const N01 = 0x01n;
const N7F = 0x7fn;
const N80 = 0x80n;
```

**Comment by samchon** [comment]
image

https://bundlephobia.com/package/typia@6.0.0-dev.20240118

Succeeded to adjust the require function calling. 

Bundling size dramatically reduced, and it would be suitable for the tree shaking.

I'll publish your PR as v5.3.12 version after detailed testing, and require statement version to be v6.

Thanks for contribution.

File: src/functional/$ProtobufReader.ts (hunk 2 above)
Comment (EloB) [reply src/functional/$ProtobufReader.ts:193]: The syntax for `BigInt` without using `BigInt(value)` is to add an "n" (`1n === BigInt(1)`) in the end of a integer.

File: src/transformers/CallExpressionTransformer.ts [hunk 3]
```diff
@@ -92,38 +92,38 @@ import { ReflectMetadataTransformer } from "./features/reflect/ReflectMetadataTr
...
     (project: IProject) =>
-    (expression: ts.CallExpression): ts.Expression | null => {
-      //----
-      // VALIDATIONS
-      //----
-      // SIGNATURE DECLARATION
-      const declaration: ts.Declaration | undefined =
-        project.checker.getResolvedSignature(expression)?.declaration;
-      if (!declaration) return expression;
+      (expression: ts.CallExpression): ts.Expression | null => {
```
Comment (EloB) [inline src/transformers/CallExpressionTransformer.ts]: I will revert these. My editor must had formatted them.

**Comment by EloB** [comment]
What do you think to extract the helpers into a separate library? What I read in some of the issues were that you couldn't have ESM output because `ts-patch` had some issues?

According to this comment.
https://github.com/samchon/typia/issues/752#issuecomment-1683044268

Is the `ts-patch` only used for the `transformer`? If so then we could have two seperate packages like `typia` (dependency) containing the helper functions and `typia-plugin-transformer` (devDependency). Then production builds only contains the helpers and not the actual compiler plugin. Making the actual npm bundle smaller :)

One thing that I didn't tell was that I use a ESM output from `typia` but only the helpers. That is the last thing I needed to get dead code elimination to work.

**Comment by EloB** [comment]
This is my draft for esbuild plugin that intercept typescript files containing strings of `"typia"` or `'typia'`. It transpiles them using the normal typescript compiler then passes the result to esbuild so it could do the dead code elimination.

It probably needs some more love before I would consider it stable but it works.

```ts
import * as ts from "typescript";
import * as fs from "fs/promises";
import * as path from "path";

const typiaEsbuildPlugin = (
  {
    tsconfigPath,
    cwd = process.cwd(),
    nodeModulesPath = path.join(cwd, "node_modules"),
  }: TypiaPluginOptions = {}
): esbuild.Plugin => ({
  name: "typia",
  setup(build) {
    let loadedConfig = {};
    let baseDir = cwd;
    const fileName = ts.findConfigFile(
      cwd,
      ts.sys.fileExists,
      tsconfigPath
    );
    if (tsconfigPath !== undefined && !fileName) throw new Error(`Failed to open '${fileName}'`);
    if (fileName) {
      const text = ts.sys.readFile(fileName);
      if (text === undefined) throw new Error(`failed to read '${fileName}'`);

      const result = ts.parseConfigFileTextToJson(fileName, text);

      if (result.error !== undefined) throw new Error(`failed to parse '${fileName}'`);

      loadedConfig = result.config;
      baseDir = path.dirname(fileName);
    }
    const parsedConfig = ts.parseJsonConfigFileContent(
      loadedConfig,
      ts.sys,
      baseDir
    );
    
    // Use the TypeScript files that are still ESM. To improve dead code elimination.
    build.onLoad({ filter: /node_modules\/typia\/lib\/index\.js$/ }, async (args) => {
      const fileName = path.join(nodeModulesPath, "typia", "src", "index.ts");
      const contents = await fs.readFile(fileName, "utf8");
      return { contents, loader: "ts", resolveDir: path.dirname(fileName) };
    });

    // Look for strings in TypeScript containging "typia" or 'typia' and compile them.
    build.onLoad({ filter: /\.([cm]?ts|tsx)$/, namespace: "file" }, async (args) => {
      const source = await fs.readFile(args.path, "utf8");
      if (!/(["'])typia\1/.test(source)) return undefined;
      const program = ts.createProgram([args.path], parsedConfig.options);
      const sourceFile = program.getSourceFile(args.path);
      let contents;
      program.emit(sourceFile, (fileName, compiledSource) => {
        contents = compiledSource
      });
      return { contents, loader: "js" };
    });
  },
});
```

**Comment by samchon** [comment]
If above plugin code works, it would be nice. Thanks for research.

It is okay to publiish by your own package like `typia-esbuild-plugin`, and also okay to sending a PR to making a namespaced module `@typia/esbuild-plugin` (or better name) into the `packages` directory.

> If make your own package, please tell me the setup way or send PR for the guide documents
>
> https://typia.io/docs/setup/

Also in the `esbuild` plugin case, no need to setup the `ts-patch`, so I think ESM problem would not be happened.

**Comment by EloB** [comment]
Yes the plugin works. There might be some issues in watch mode but I will take a look at that. For instance if the "type" that is consumed by Typia is written in another file. I don't think it will rebuild then because it isn't watched right now but the plugin system has ways to tell what files to listen to.

Like this:
```ts
import typia from "typia";
import type { MyResponse } from "./types";

const validate = typia.createIs<MyResponse>();
```

**Comment by EloB** [comment]
It doesn't include the source maps right but I will fix that too.

**Comment by samchon** [comment]
```typescript
const createValidateDecodePure = /** @__PURE__ */ Object.assign<
  typeof createValidateDecode,
  {},
  {}
>(
  createValidateDecode,
  /** @__PURE__ */ Namespace.validate(),
  /** @__PURE__ */ Namespace.protobuf.decode("createValidateDecode"),
);
export { createValidateDecodePure as createValidateDecode };
```
 
@EloB Changed like above to hide namespace nested functions in the declaration files.

It seems all things are ready and if you agree, I'll publish it.

**Review by samchon — APPROVED** [review]
Good, I'd missed somethinig.

Now, may I do publish?

**Review by samchon — APPROVED** [review]
Everything working fine. Thanks for contribution and waiting for your esbuild plugin.

**Comment by EloB** [comment]
@samchon Thanks mate for such fast feedback, reviewing and merging. I don't know if you noticed my last push/commit there. I made a change to use the correct reference in the `Object.assign`. The dead code elimination in `esbuild` were to aggressive and were removing code that shouldn't be removed. The last commit corrected that so dead code elimination works as expected now.

Also want to thank you for this amazing tool/plugin/library! I'm really impressed what you have achieved.

# Another interesting find
I discovered it's better to use `typia` like this if you want proper dead code elimination.
```ts
// Good, allowed proper dead code elimination.
import { is, tags } from "typia";

// Good, allowed proper dead code elimination.
import * as typia from "typia";

// Bad, was including everything in my bundle. 
import typia from "typia";

// Bad, was including everything in my bundle. 
import typia, { is, tags } from "typia";
```

# Still on my wishlist
I also want typia as ESM output. That is one of the last thing that is needed for dead code elimination.

**Pull request merged**